### PR TITLE
feat(2135): New endpoint to create and validate pipeline template configuration

### PR DESCRIPTION
**Pull request by klu909**
## Context

replicate of https://github.com/screwdriver-cd/screwdriver/pull/2900
## Objective

add new endpoints to create and validate pipeline template configuration
## References

https://github.com/screwdriver-cd/screwdriver/issues/2135
## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by klu909** [comment]
this pr is blocked by https://github.com/screwdriver-cd/template-validator/pull/33, test will fail until it's merged

**Comment by klu909** [comment]
not going to add functional test in this pr, that should be done after all new endpoint has been added

File: plugins/pipelineTemplates/create.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+const urlLib = require('url');
+const boom = require('@hapi/boom');
+const schema = require('screwdriver-data-schema');
+const validator = require('screwdriver-template-validator').parsePipelineTemplate;
+const templateSchema = schema.api.templateValidator;
+
+module.exports = () => ({
+    method: 'POST',
+    path: '/pipelineTemplates',
```
Comment (pritamstyz4ever) [inline plugins/pipelineTemplates/create.js:11]: should we use /pipeline/tempalte instead and move the files under the pipeline folder to maintain the natural structure of the project

File: plugins/pipelineTemplates/create.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    options: {
+        description: 'Create a new pipeline template',
+        notes: 'Create a specific pipeline template',
+        tags: ['api', 'pipelineTemplate'],
+        auth: {
+            strategies: ['token'],
+            scope: ['build']
+        },
+
+        handler: (request, h) => {
```
Comment (pritamstyz4ever) [inline plugins/pipelineTemplates/create.js:21]:
```suggestion
        handler: (request, h) => {
                    const { pipelineTemplateVersionFactory, pipelineTemplateFactory } = request.server.app;

              const config = await validator(request.payload.yaml);
  
              if (config.errors.length > 0) {
                  throw boom.badRequest(`Template has invalid format: ${config.errors.length} error(s).`, config.errors);
              }
  
              const pipelineTemplate = await pipelineTemplateFactory.get({
                  name: config.template.name,
                  namespace: config.template.namespace
              });
              const { pipelineId } = request.auth.credentials;
  
              if (pipelineTemplate) {
                  if (pipelineTemplate.pipelineId !== pipelineId) {
                      throw boom.notFound(`Pipeline IDs do not match`);
                  }
              }
  
              const template = await pipelineTemplateVersionFactory.create(
                  {
                      ...config.template,
                      pipelineId
                  },
                  pipelineTemplateFactory
              );
  
              const urlObject = {
                  pathname: `${request.path}/${template.id}`
              };
              const location = new URL(
                  urlObject,
                  `${request.server.info.protocol}://${request.headers.host}:${request.headers.port}`
              ).toString();
  
              return h.response(template.toJson()).header('Location', location).code(201);
        }
```

File: plugins/pipelineTemplates/create.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                                {
+                                    ...config.template,
+                                    pipelineId
+                                },
+                                pipelineTemplateFactory
+                            );
+                        });
+                })
+                .then(template => {
+                    const location = urlLib.format({
```
Comment (pritamstyz4ever) [inline plugins/pipelineTemplates/create.js:57]: urlLib.format is deprecated

**Review by pritamstyz4ever — CHANGES_REQUESTED** [review]

File: plugins/pipelineTemplates/create.js (hunk 2 above)
Comment (klu909) [reply plugins/pipelineTemplates/create.js:21]: https://hapi.dev/api/?v=21.3.2#-requestheaders `request.headers.port` appears to be invalid and has been covered up by this urlLib.format function

**Comment by klu909** [comment]
https://hapi.dev/api/?v=21.3.2#-requestheaders request.headers.port appears to be invalid and has been covered up by this urlLib.format function

File: plugins/pipelineTemplates/create.js (hunk 3 above)
Comment (pritamstyz4ever) [reply plugins/pipelineTemplates/create.js]: maybe we don't need the port, It should work without it

File: plugins/pipeline-template-validator.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * @method register
+ * @param  {Hapi.Server}    server
+ * @param  {Function} next
+ */
+const pipelineTemplateValidatorPlugin = {
+    name: 'pipeline-template-validator',
+    async register(server) {
+        server.route({
+            method: 'POST',
+            path: '/validator/pipelineTemplate',
```
Comment (pritamstyz4ever) [inline plugins/pipeline-template-validator.js:21]: this should be changed to
```suggestion
            path: '/pipeline/template/validate',
```

Comment (pritamstyz4ever) [reply plugins/pipeline-template-validator.js:21]: Move the API under  pipelines/templates/

**Review by pritamstyz4ever — CHANGES_REQUESTED** [review]

File: lib/server.js [hunk 5]
```diff
@@ -126,6 +126,11 @@ module.exports = async config => {
             commandTagFactory: config.commandTagFactory,
             templateFactory: config.templateFactory,
             templateTagFactory: config.templateTagFactory,
+            pipelineTemplateFactory: config.pipelineTemplateFactory,
+            pipelineTemplateVersionFactory: config.pipelineTemplateVersionFactory,
```
Comment (pritamstyz4ever) [inline lib/server.js:130]:
```suggestion
            pipelineTemplateVersionFactory: config.pipelineTemplateVersionFactory,
```

**Review by pritamstyz4ever — COMMENTED** [review]
LGTM

**Review by pritamstyz4ever — APPROVED** [review]

**Pull request merged**